### PR TITLE
Add Ordos 6a

### PR DESCRIPTION
**Pull request by abcdefg30**
The highly awaited sequel to #16057 - exactly one year after the original. 🎉 

I slightly increased the time between waves on the hardest difficulty, making it beatable without requiring excessive effort (I can easily beat it before the reinforcements arrive). Additionally I decreased the waiting time for the reinforcements. 30 to 40 minutes are just really long, and most of the time you'll have completed the mission already by that time. The maximum wait time is now 20 minutes.

Other changes include:
- A display for when the next reinforcements arrive
- You can no longer sell the starport (the one you have to "guard at all costs") and still win
- Code refactors

### CODE REVIEW

**Comment by MustaphaTR** [comment]
In the original version of the mission, you didn't have to guard that starport specifically. You could build a new Starport in the main base and get rid of the other Starport to be able to defend it better.

**Comment by abcdefg30** [comment]
Figured as much. I'd not be surprised if this was actually a bug in the original.
Implementing this posed a slight problem: How was the player going to know about this?
The primary objective is very concise and I didn't want to convolute it. Additionally, I don't think that makes the mission better. Allowing this would just turn the mission into yet another "protect your base and destroy the enemy base" mission.

File: OpenRA.sln [hunk 1]
```diff
@@ -201,7 +203,6 @@ Global
 		{F33337BE-CB69-4B24-850F-07D23E408DDF}.Debug|Any CPU.Build.0 = Debug|Any CPU
 		{F33337BE-CB69-4B24-850F-07D23E408DDF}.Release|Any CPU.ActiveCfg = Release|Any CPU
 		{F33337BE-CB69-4B24-850F-07D23E408DDF}.Release|Any CPU.Build.0 = Release|Any CPU
-		{F33337BE-CB69-4B24-850F-07D23E408DDF}.Debug|Any CPU.Build.0 = Debug|Any CPU
```
Comment (fruestueck) [inline OpenRA.sln:204]: Not sure if this is intentional.

**Review by fruestueck — DISMISSED** [review]
At my first playthrough attempt, an enemy tank stood idle for some time next to the starport. That was helpful.
[OpenRA-2018-01-02T033026Z.zip](https://github.com/OpenRA/OpenRA/files/4097016/OpenRA-2018-01-02T033026Z.zip)

Haven't been able to beat the mission with four attempts on default difficulty easy. For an unskilled player like me, who would enjoy missions and story over hardcore gaming, some "baby" difficulty would be nice. But I'm not the typical dune2k player and don't represent the community playing it.

Was able to win just with debug commands (on my second try with debug *shame*) :D

File: OpenRA.sln (hunk 1 above)
Comment (abcdefg30) [reply OpenRA.sln:204]: It is, see the first commit (message).

**Comment by abcdefg30** [comment]
Hm, ok. I'll look into making it more easy. (Fwiw, I used the standard OpenRA-d2k method of just building 438 rifle infantry and swarming the map.)

**Comment by fruestueck** [comment]
> Hm, ok. I'll look into making it more easy. (Fwiw, I used the standard OpenRA-d2k method of just building 438 rifle infantry and swarming the map.)

I wasn't aware of that and build tech asap. Thanks for the advice.

**Comment by abcdefg30** [comment]
Rebased (first push) and updated (second). The mission should be easier now.

File: mods/d2k/maps/ordos-06a/ordos06a.lua [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+SendContraband = function()
+	Media.PlaySpeechNotification(player, "Reinforce")
+
+	for i = 0, 6 do
+		local c = player.Color
+		if i % 2 == 0 then
+			c = HSLColor.White
+		end
+
+		Trigger.AfterDelay(DateTime.Seconds(i), function() UserInterface.SetMissionText("Ixian reinforcements have arrived!", c) end)
```
Comment (pchote) [inline mods/d2k/maps/ordos-06a/ordos06a.lua:153]: Do we need an API for blinking mission text?

**Review by pchote — COMMENTED** [review]
I didn't notice any real showstoppers, although I'd argue that the difficulty just about counts.

https://www.youtube.com/watch?v=lHpn403jSkI appears to show some things that aren't in this version:
* Objective is "Defend the Starport" rather than "Guard the Starport at all costs"
* "Enemy reinforcements have arrived" notification.
* "Deliveries beginning to arrive. Massive reinforcements expected!" message is missing.

Other issues I noticed while playing:
* This is still really hard - even on easy it took me several attempts and IMO a lot more work and less fun than deserved on easy.
* A Harkonnen carryall didn't die when I defeated them, and stayed endlessly circling on the map.
* The wording "First batch of reinforcements" doesn't feel polished - I like that we show this instead of a context-less timer, but I would change this to "Initial reinforcements" instead.

**Comment by pchote** [comment]
re difficulty: the biggest issue is the fog of war, which gives no time to position forces to defend.

According to https://github.com/OpenRA/OpenRA/pull/16057#issuecomment-497387913 the enemy reinforcements notification would also help.

**Comment by reaperrr** [comment]
My 2 cents: I still think it's ridiculous (not just in this mod and mission) that our 'Easy' is usually harder than at least 'Normal', and probably even 'Hard' in the original games. In my opinion it's just wrong to base our difficulty on the assumption that the original games were too easy for most and that 99% of players are seasoned veterans who would get bored by something closer to the original games, because I think neither is accurate.

Though I think the easiest solution at this point might be to just rename our current difficulties to Normal - Hard - Very Hard/Brutal/Nightmare, and add a _real_ Easy difficulty below that.

File: mods/d2k/maps/ordos-06a/ordos06a.lua (hunk 2 above)
Comment (abcdefg30) [reply mods/d2k/maps/ordos-06a/ordos06a.lua:153]: Might be a good idea, since we use this pattern more often. But not in this PR.

**Comment by abcdefg30** [comment]
Rebased and updated.

**Review by pchote — APPROVED** [review]

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/2e307e8e3ba7842e29877582a29ea1e1d9326c3d...1cf6a0bdbb62ee0665a3afe3fca1c353ef6a0872

2a26a07
d0d55b3
a01233c
8690ad4

**Pull request merged**